src/os/bluestore/BlueStore.cc

```diff
@@ -8108,18 +8118,22 @@ void BlueStore::_txc_finish(TransContext *txc)
   }
 }
 
-void BlueStore::_txc_release_alloc(TransContext *txc)
+void BlueStore::_txc_release_alloc_locked(TransContext *txc)
 {
-  // update allocator with full released set
+  // it's expected we're called with lazy_release_lock already taken!
   if (!cct->_conf->bluestore_debug_no_reuse_blocks) {
     dout(10) << __func__ << " " << txc << " " << txc->released << dendl;
-    for (interval_set<uint64_t>::iterator p = txc->released.begin();
-	 p != txc->released.end();
-	 ++p) {
-      alloc->release(p.get_start(), p.get_len());
-    }
+    // interval_set seems to be too costly for inserting things in
+    // bstore_kv_final. We could serialize in simpler format and perform
+    // the merge separately, maybe even in a dedicated thread.
+    lazy_release_extents.insert(txc->released);
   }
 
+  if (lazy_release_extents.num_intervals() >= 256) {
+    // the remaining part will be cleaned in _close_alloc().
+    alloc->release(lazy_release_extents);
+    lazy_release_extents.clear();
```

Comment:
it's short, but i would still put this in a helper like [_]release_lazy_extents() that can be called from here or elsewhere

Reply:
Then, in the periodic cache trim thread, which wakes up a few times a second, add the conditional release.  I would set the aggressive_release bool to true if the free space is below 5% or similar. 

For the non-aggressive release condition, it would be nice if it actually counted the # of blocks, not extents; we don't know if this is gigabytes or kilobytes of space.  Until then, the num_intervals() is fine.

Reply:
Sure. TODO.